Split OpenFL/lime command 'test' into independent 'update, build' and 'run' commands.

EricBishton opened the pull request.

EricBishton commented:
The big thing here is that HaxeCommonCompilerUtil.java was refactored.  It will probably be easier to read from scratch, rather than try to understand the diff.  Much of the functionality was moved around in the file, so diff is fairly confused.  😕

EricBishton commented:
Hmmm.  I see Codacy needs to be tuned:
 - I don't really care about non-optimized imports.  They get cleaned up eventually.  Everybody tries.
 - Ordering of private classes and fields is fully up to the dev and what makes sense for the layout of the code.  Arbitrary rules about where they need to appear don't help.
 - It's fine to not combine nested if statements when they show different levels/types of logic.  Combining if statements just because they _can_ be combined does not good code make.  Plus there is no performance benefit to doing so.

mayakwd commented:
- In my opinion ordering of private fields have sense. I always expect that constants and final variables are located at the beginning of the class. Actually I prefer to have all private fields declared at the beginning. But anyways you could use your own layout.
- The arrangement of methods in the order of calling (at least approximately) adds 100 points to code readability.
- About switch statements - I believe that `default` case with thrown `unimplemented` `exception reduce a lot of work in the future. 
- Unused imports - is just a junk in the code (you could set "optimize imports before commit" in the commit dialog. But it's not necessary.

mayakwd commented on common/src/com/intellij/plugins/haxe/util/HaxeCommonCompilerUtil.java:
```diff
@@ -95,155 +100,229 @@ public static boolean compile(final CompilationContext context) {
+    }
+    final File workingDirectory = new File(FileUtil.toSystemDependentName(workingPath));
+    if (!workingDirectory.exists()) {
+      if (!workingDirectory.mkdirs()) {
```
Well, these if-conditions really could be combined, but I understand your embarrassing.  There is some "inexplicable prejudice"  about short-circuit evaluation. Developers prefer not to combine statements with action's result in one condition, because we do not fully believe if the action will not be executed if statement equals "false".

EricBishton replied:
Well, I'm not embarrassed.  And it's not about short-circuit evaluation, either.  I didn't combine them because line 113 is basically a test, and line 114 is some work that is to be performed, that just happens to return a boolean that we can test.  Certainly, I can combine them, but it doesn't read as well to me, -- or make better code, in my opinion --  and that's why I didn't.  

Perhaps I would be truer to my own philosophy here if I captured the boolean result and tested that with a separate if statement.

mayakwd commented on common/src/com/intellij/plugins/haxe/module/impl/HaxeModuleSettingsBaseImpl.java:
```diff
@@ -200,19 +200,19 @@ public void setHxmlPath(String hxmlPath) {
 
   public boolean isUseHxmlToBuild() {
-    return buildConfig == USE_HXML;
+    return HaxeConfiguration.translateBuildConfig(buildConfig) == HaxeConfiguration.HXML;
```
Could you make property to reduce code duplication? There is a 5 similar places in this diff.

EricBishton replied:
What are you expecting?  `return HaxeConfiguration.HXML.asBuildConfig() == buildConfig;`?  Or do you have something else in mind?

EricBishton replied:
Or, do you mean a _method_ instead of _property_?

mayakwd replied:
Yes, method:
```java
private boolean isConfigurationKindOf(HaxeConfiguration config) {
   return HaxeConfiguration.translateBuildConfig(buildConfig) == config;
}
```

mayakwd commented on common/src/com/intellij/plugins/haxe/util/HaxeCommonCompilerUtil.java:
```diff
@@ -95,155 +100,229 @@ public static boolean compile(final CompilationContext context) {
+    try {
+      for (List<String> commandLine : commandLines) {
+
+        // XXX: Would it be useful to show the working directory in the output window??
```
And output directory and full command line with arguments.

EricBishton replied:
I'm not sure that I have the output directory available.  That's up to haxe (or lime) and while I have some heuristics that seem to hold, I'm not sure what the compiler will actually do.  But I'll put out the extra info that I have.

mayakwd commented on common/src/com/intellij/plugins/haxe/util/HaxeCommonCompilerUtil.java:
```diff
@@ -95,155 +100,229 @@ public static boolean compile(final CompilationContext context) {
-      if (mainClass == null || mainClass.length() == 0) {
-        context.errorHandler(HaxeCommonBundle.message("no.main.class.for.module", context.getModuleName()));
+
+      if ( !verifyNonEmptyString(mainClass, context, "no.main.class.for.module")
```
How do you think - is it be better to show user all problems at once, not one by one?

EricBishton replied:
That's an interesting point.  I hadn't thought about it because I was just repackaging the original code.  Rather than short-circuit, I'll put out the error messages and remember that there was an error, then return appropriately at the end.

mayakwd replied:
That would be great!

EricBishton commented:
> I always expect that constants and final variables are located at the beginning of the class.

In general, when they are used in multiple places, or they are public and other classes are likely to use them, I agree.

In some cases, we use a constant (or, really, just a static final field) as an optimization for a value that we'd rather not recompute.  When they are private and appear in only one method, it makes more sense to have the field accompany the method which uses it, rather than be at the top of the (possibly large) class.

EricBishton commented:
My comment about Codacy isn't that I bridle at the prospect of having my code reviewed automatically.  It's that the tool is attempting to enforce rules which I don't agree with.  Thus, the comment about tuning it.  You'll notice that there were several warnings that it issued that I didn't comment on, and it did find some interesting things that I am thinking about.  

That said, until we agree on a set of rules for it to follow, I'm not going to give it any credence.

mayakwd commented:
I've turned off codacy review and reverted badge.

EricBishton commented:
Thank you for reverting the badge.  I want to have a conversation with you about tuning it.  I think it might be a useful tool and I'm willing to give it a try if we can come to agreement.

mayakwd commented:
Review is done, the only change I want to see error messages should be all in one.

EricBishton commented:
Well, I've got a lot more than that.  I fixed the repeated code issue that you pointed out, and I've fixed the missing default case.  But testing also led me down a path to fixing error parsing, too.
You'll see it in another hour, probably.  I'm submitting changes now.
